reactor-netty-http/src/main/java/reactor/netty/http/server/NonSslRedirectDetector.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2011-Present VMware, Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.netty.http.server;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.socket.ChannelInputShutdownEvent;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.ssl.SslHandler;
+import reactor.netty.NettyPipeline;
+import reactor.netty.tcp.SslProvider;
+
+import javax.annotation.Nullable;
+import java.net.SocketAddress;
+
+import static io.netty.handler.codec.ByteToMessageDecoder.MERGE_CUMULATOR;
+
+/**
+ * @author James Chen
+ */
+final class NonSslRedirectDetector extends ChannelInboundHandlerAdapter {
```

Comment:
Can we extend `io.netty.handler.ssl.OptionalSslHandler`?

Reply:
No. According to the implementation of OptionalSslHandler, it can only replace itself with another handler for non-ssl data or ssl data. But `sslProvider.addSslHandler(ctx.channel(), remoteAddress, sslDebug);` will add multiple handlers in `NonSslRedirectDetector#handlePotentialHeader` (and then NonSslRedirectHandler will just remove itself).

Reply:
Yes true.
Why don't we use the approach that `OptionalSslHandler` uses and extend `ByteToMessageDecoder`? Thus it will handle for us the cumulation work.

Reply:
Just for code clarity, it's eaiser to read `NonSslRedirectDetector` + `ChannelInboundHandlerAdapter` instead of `NonSslRedirectDetector` + `ByteToMessageDecoder` + `ChannelInboundHandlerAdapter`

e.g. The code in channelRead() of `NonSslRedirectDetector` is far easier to read than the one in ByteToMessageDecoder
https://github.com/reactor/reactor-netty/blob/31664c33b2e78d26b54b6be8e7979b5bec25762d/reactor-netty-http/src/main/java/reactor/netty/http/server/NonSslRedirectDetector.java#L57-L68

Reply:
Agree, on the other hand, having `NonSslRedirectDetector` as an extension of `ByteToMessageDecoder` gives us the ability to focus only on the functionality that this handler needs to do and leave `ByteToMessageDecoder` to handle various use cases when the cumulation has to be released.
Wdyt?

Reply:
You can try to implement the same functions by extending `ByteToMessageDecoder` locally and you may understand what I mean:

> it's far eaiser to read `NonSslRedirectDetector` + `ChannelInboundHandlerAdapter` instead of `NonSslRedirectDetector` + `ByteToMessageDecoder` + `ChannelInboundHandlerAdapter`

The code extends from `ByteToMessageDecoder` is definitely shorter, but definitely harder to debug and read (bad maintainability).

Reply:
I was looking at `OptionalSslHandler`. Do we need some more functionality that what `OptionalSslHandler` does?

Reply:
Yes, we need more functionality than what `OptionalSslHandler` can provide.
For the obvious one, we need to call `sslProvider.addSslHandler(ctx.channel(), remoteAddress, sslDebug);` that will add multiple SSL-related handlers in `NonSslRedirectDetector`, rather than just `context.pipeline().replace(this, newSslHandlerName(), sslHandler);` in `OptionalSslHandler`

Reply:
I'm sorry I didn't explain correctly. I was thinking for something like this:

```
    @Override
    protected void decode(ChannelHandlerContext context, ByteBuf in, List<Object> out) throws Exception {
        if (in.readableBytes() < SslUtils.SSL_RECORD_HEADER_LENGTH) {
            return;
        }
		if (SslHandler.isEncrypted(buf)) {
			// Lazy init for better performance
			sslProvider.addSslHandler(ctx.channel(), remoteAddress, sslDebug);
			ChannelHandlerContext context = pipeline.context(NettyPipeline.SslHandler);
			SslHandler sslHandler = (SslHandler) context.handler();
			sslHandler.channelRead(context, buf);
		}
		else {
			pipeline.addBefore(NettyPipeline.ReactiveBridge, NettyPipeline.NonSslRedirectHandler, new NonSslRedirectHandler());
			ctx.fireChannelRead(buf);
		}
    }
```

Wdyt?

Reply:
Our new tests will fail because of `Caused by: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1` with the code. Full code:

```
final class NonSslRedirectDetector extends ByteToMessageDecoder {

	private static final int SSL_RECORD_HEADER_LENGTH = 5;

	private final SslProvider sslProvider;
	private final SocketAddress remoteAddress;
	private final boolean sslDebug;

	public NonSslRedirectDetector(SslProvider sslProvider, @Nullable SocketAddress remoteAddress, boolean sslDebug) {
		this.sslProvider = sslProvider;
		this.remoteAddress = remoteAddress;
		this.sslDebug = sslDebug;
	}

	@Override
	protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
		if (in.readableBytes() < SSL_RECORD_HEADER_LENGTH) {
			return;
		}

		ChannelPipeline pipeline = ctx.pipeline();
		pipeline.remove(this);
		if (SslHandler.isEncrypted(in)) {
			// Lazy init for better performance
			sslProvider.addSslHandler(ctx.channel(), remoteAddress, sslDebug);
			ChannelHandlerContext context = pipeline.context(NettyPipeline.SslHandler);
			SslHandler sslHandler = (SslHandler) context.handler();
			sslHandler.channelRead(context, in);
		}
		else {
			pipeline.addBefore(NettyPipeline.ReactiveBridge, NettyPipeline.NonSslRedirectHandler, new NonSslRedirectHandler());
			ctx.fireChannelRead(in);
		}
	}

}
```

Caused by: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:74)
	at io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:138)
	at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:100)
	at io.netty.buffer.WrappedByteBuf.release(WrappedByteBuf.java:1037)
	at io.netty.buffer.SimpleLeakAwareByteBuf.release(SimpleLeakAwareByteBuf.java:102)
	at io.netty.buffer.AdvancedLeakAwareByteBuf.release(AdvancedLeakAwareByteBuf.java:941)
	at io.netty.handler.codec.ByteToMessageDecoder.handlerRemoved(ByteToMessageDecoder.java:256)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:515)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:447)

Reply:
what I have to replace in the PR in order to see this exception? just the detector?

Reply:
Yes. Just need to replace `NonSslRedirectDetector` with the code. And the exception should occur when running `HttpRedirectTests`

Reply:
Ok try with

```
	@Override
	protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
		if (in.readableBytes() < SSL_RECORD_HEADER_LENGTH) {
			return;
		}

		ChannelPipeline pipeline = ctx.pipeline();
		if (SslHandler.isEncrypted(in)) {
			sslProvider.addSslHandler(ctx.channel(), remoteAddress, sslDebug);
		}
		else {
			pipeline.addBefore(NettyPipeline.ReactiveBridge, NettyPipeline.NonSslRedirectHandler, new NonSslRedirectHandler());
		}
		pipeline.remove(this);
	}
```

and in the SslProvider#addSslHandler

```
		ChannelPipeline pipeline = channel.pipeline();
		if (pipeline.get(NettyPipeline.ProxyHandler) != null) {
			pipeline.addAfter(NettyPipeline.ProxyHandler, NettyPipeline.SslHandler, sslHandler);
		}
+		else if (pipeline.get(NettyPipeline.NonSslRedirectDetector) != null) {
+			pipeline.addAfter(NettyPipeline.NonSslRedirectDetector, NettyPipeline.SslHandler, sslHandler);
+		}
		else {
			pipeline.addFirst(NettyPipeline.SslHandler, sslHandler);
		}
```

and in the SniProvider#addSniHandler

```
		if (pipeline.get(NettyPipeline.NonSslRedirectDetector) != null) {
			pipeline.addAfter(NettyPipeline.NonSslRedirectDetector, NettyPipeline.SslHandler, newSniHandler());
		}
		else {
			pipeline.addFirst(NettyPipeline.SslHandler, newSniHandler());
		}
```

Reply:
updated. ptal.